Update module structure and crate docs

CPerezz opened the pull request:
Since `v0.1.0` contained errors in the implementation and didn't take care about the library structure. This PR fixes all of the tech-debt that we had there. 

Bump to `v0.2.0`

Commit:
Refactor lib module structure

Commit:
Add CONTRIBUTING file

Commit:
Update README.md

Commit:
Refactor module-structure of the crate

Commit:
Include lints for the lib

Commit:
Fix clippy lints

Commit:
Add kev to authors

Commit:
Bump to v0.2.0

Commit:
Add clippy checks in travis builds.

Commit:
Add auto-doc uploads

Commit:
Export `AllocatedScalar`

Commit:
Derive `Copy` for `AllocatedScalar`

vlopes11 commented on src/gadgets/scalar_gadgets/mod.rs:
```diff
@@ -0,0 +1,24 @@
+pub struct AllocatedScalar {
+    /// Variable associated to the `Scalar`.
+    pub var: Variable,
+    /// Scalar associated to the `Varaible`
```
Small typo

CPerezz replied:
Fixed. Ty!

vlopes11 reviewed (approved):
LGTM, with a small typo

Commit:
Fix typo

Commit:
Use dusk emails in Cargo.toml authors section

ZER0 commented on src/lib.rs:
```diff
@@ -1 +1,40 @@
+#![allow(clippy::many_single_char_names)]
+// Bool expr are usually easier to read with match statements.
+#![allow(clippy::match_bool)]
+#![deny(intra_doc_link_resolution_failure)]
```
replace `intra_doc_link_resolution_failure` with `broken_intra_doc_links`.

CPerezz replied:
Done.

ZER0 commented on src/lib.rs:
```diff
@@ -1 +1,40 @@
+// Variables have always the same names in respect to wires.
+#![allow(clippy::many_single_char_names)]
+// Bool expr are usually easier to read with match statements.
+#![allow(clippy::match_bool)]
```
Remove this.
First of all, there is no such code in the repo that needs this rule.

Second of all, the comment is highly opinionated, and we should comment fact not opinion.

In our code, we use both `if` and `match` statements and expressions, since `if` is a rust sugared syntax for `match`ing boolean in Rust.

Of course we have always to evaluate case by case, but unless explicitly said otherwise for matching a boolean we usually use `if` instead of `match` (both for statement and expression).

CPerezz replied:
Right. As we discussed it's removed.

ZER0 commented on src/lib.rs:
```diff
@@ -1 +1,40 @@
+#![doc(html_favicon_url = "https://dusk.network/lib/img/favicon-16x16.png")]
+#![allow(clippy::suspicious_arithmetic_impl)]
+// Some structs do not have AddAssign or MulAssign impl.
+#![allow(clippy::suspicious_op_assign_impl)]
```
I think we don't have any code in this repo that needs them, double check and if that is the case please remove them.

CPerezz replied:
I left the `suspicious_arithmetic_impl` since we have bitshift ops.

ZER0 commented on src/gadgets/scalar_gadgets/range.rs:
```diff
@@ -120,69 +110,26 @@ pub fn max_bound(
 
     // Take the first n bits
-    let scalar_bits_var = scalar_bits_var[0..num_bits].to_vec().clone();
+    let scalar_bits_var = scalar_bits_var[0..num_bits].to_vec();
```
Nit:`0` is superfluous.

CPerezz replied:
Fixed

ZER0 commented on src/gadgets/scalar_gadgets/range.rs:
```diff
@@ -192,20 +139,15 @@ fn scalar_decomposition_gadget(
-
         composer.boolean_gate(*bit);
 
         let two_pow = BlsScalar::from(2).pow(&[power as u64, 0, 0, 0]);
```
Did we land the `pow_of_2` yet?

CPerezz replied:
We did. I'll add it.

ZER0 commented on src/gadgets/scalar_gadgets/range.rs:
```diff
@@ -350,7 +291,7 @@ mod tests {
 
     #[test]
-    fn max_bound_test()  {
+    fn max_bound_test() {
```
You can make the tests returns a `Result` instead of using `expect`.

CPerezz replied:
Done.

ZER0 reviewed (changes requested):
There are small things to change in this PR, but we might take the occasion to actually doing a refactoring on the module structure (since this seems the topic as well), plus there are few piece of code in the repo that needs a refactor (but weren't touched by this PR)

Also the tests, it's not totally clear to me (at least at first sight) if they're integration tests or not, so maybe that's another thing to take into account.

Commit:
Address all of @ZERO 's nits in #18

Commit:
Move integration tests to a test folder

As correctly suggested by @ZER0 we should move the integration
tests to a `tests` folder next to `src` so cargo does compile and
run them.

That allows us to see how the API looks like and therefore prevent
us to push anythig that does not look good related to the API shape.

CPerezz commented:
Integration tests added @ZER0. Could you review this again?

Commit:
Address clippy lints

ZER0 reviewed (approved):
LGTM
